### PR TITLE
Just another CI update for easier maintain

### PR DESCRIPTION
**Pull request by martadinata666**
As the building binary pretty much stable, we dont need separate 3 job target. This PR merge 3 job into one by using matrix
* https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs
* only binary building affected
* add minimal postgres and mysql integration (up and run), selfcheck
* Selfcheck will throw err if something bad happen

### CODE REVIEW

File: .github/workflows/docker-build-static.yml [hunk 1]
```diff
@@ -151,82 +112,90 @@ jobs:
...
+       - name: Download All Artifacts
+         uses: actions/download-artifact@v3
+         with:
+           path: bin/
+       - name: Set executables
+         run: chmod +x bin/*/lldap
+       - name: run lldap with postgres DB
+         run: |
+              bin/x86_64-unknown-linux-musl-lldap-bin/lldap run &
+              sleep 10s
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:177]: Ha! That's one way to do it :D 
Does that actually work?

Comment (nitnelave) [reply .github/workflows/docker-build-static.yml:177]: If so, we should also do one with sqlite.

File: .github/workflows/docker-build-static.yml [hunk 2]
```diff
@@ -86,58 +87,18 @@ jobs:
           name: ui
           path: app/
 
-  build-armhf:
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:89]: Nice cleanup!

File: .github/workflows/docker-build-static.yml [hunk 3]
```diff
@@ -34,7 +34,7 @@ env:
 ### run app/build.sh
 ### upload artifacts
 
-# builds-armhf, build-aarch64, build-amd64 create binary for respective arch
+# builds-armhf, build-aarch64, build-amd64 create binary for respective arch look for .github/workflows/Dockerfile.dev for deps list
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:37]: Move the added comment to a new line, new sentence?

File: .github/workflows/Dockerfile.dev [hunk 4]
```diff
@@ -31,4 +31,12 @@ RUN wget -c https://musl.cc/x86_64-linux-musl-cross.tgz && \
     rm ./x86_64-linux-musl-cross.tgz && \
     rm ./aarch64-linux-musl-cross.tgz
 
+### armhf deps
```
Comment (nitnelave) [inline .github/workflows/Dockerfile.dev:34]: I'll need to rebuild/republish the build image, right?

File: .github/workflows/docker-build-static.yml (hunk 1 above)
Comment (martadinata666) [reply .github/workflows/docker-build-static.yml:177]: Well it works, and `lldap selfcheck` pretty much success. Idk if it enough, it just ensure it run fine first. ☕

Comment (nitnelave) [reply .github/workflows/docker-build-static.yml:177]: LLDAP Healthcheck doesn't actually return a non-0 code if there are errors, though. You'd need to grep for ERROR in the output (or I can make a change).

File: .github/workflows/Dockerfile.dev (hunk 4 above)
Comment (martadinata666) [reply .github/workflows/Dockerfile.dev:34]: right, im thinking keep the deps in one development image. 👍🏼

File: .github/workflows/docker-build-static.yml (hunk 1 above)
Comment (nitnelave) [reply .github/workflows/docker-build-static.yml]: Fixing it in https://github.com/nitnelave/lldap/pull/421

File: .github/workflows/Dockerfile.dev [hunk 5]
```diff
@@ -31,4 +32,12 @@ RUN wget -c https://musl.cc/x86_64-linux-musl-cross.tgz && \
...
+RUN sudo dpkg --add-architecture armhf && \
```
Comment (nitnelave) [inline .github/workflows/Dockerfile.dev:36]: Remove all the `sudo`, we're `root` and `sudo` is not installed.

File: .github/workflows/docker-build-static.yml [hunk 6]
```diff
@@ -19,35 +19,52 @@ on:
...
-### Upload artifacts
+### aarch64 and amd64 is musl based
+### armv7 is glibc based, musl had issue with time_t when cross compile https://github.com/rust-lang/libc/issues/1848
 
 ## the CARGO_ env
-#CARGO_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_LINKER: arm-linux-gnueabihf-gcc
-# This will determine which architecture lib will be used.
+### This will determine which architecture linker
+### CARGO_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_LINKER: arm-linux-gnueabihf-gcc
+### CARGO_TARGET_AARCH64_UNKNOWN_LINUX_MUSL_LINKER: aarch64-linux-musl-gcc
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:45]: No need to re-list the variables, they're also in the job definition.

File: .github/workflows/docker-build-static.yml [hunk 7]
```diff
@@ -68,76 +85,36 @@ jobs:
...
-          name: armhf-migration-tool-bin
-          path: target/armv7-unknown-linux-gnueabihf/release/migration-tool
-
 
-  build-aarch64:
+  build-bin:
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        target: [armv7-unknown-linux-gnueabihf,aarch64-unknown-linux-musl,x86_64-unknown-linux-musl]
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:111]: Nit: spaces between the values in the list.

File: .github/workflows/docker-build-static.yml [hunk 8]
```diff
@@ -151,93 +128,109 @@ jobs:
             .cargo/registry/index
             .cargo/registry/cache
             .cargo/git/db
-            target
-          key: lldap-bin-aarch64-${{ hashFiles('**/Cargo.lock') }}
+            target/
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:131]: Do we need that trailing slash? If so, can you add it above?

File: .github/workflows/docker-build-static.yml [hunk 9]
```diff
@@ -151,93 +128,109 @@ jobs:
...
-#             echo "/opt/aarch64-linux-musl-cross:/opt/aarch64-linux-musl-cross/bin" >> $GITHUB_PATH
-      - name: add musl aarch64 target
-        run: rustup target add aarch64-unknown-linux-musl
-      - name: build lldap aarch4
-        run: cargo build --target=aarch64-unknown-linux-musl --release -p lldap -p migration-tool
-      - name: check path
-        run: ls -al target/aarch64-unknown-linux-musl/release/
-      - name: upload aarch64 lldap artifacts
+            lldap-bin-${{ matrix.target }}-
+      # armhf not existed in dev image, so install here. can be removed after update dev image
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:135]: The dev image was already updated, you can remove this.

File: .github/workflows/docker-build-static.yml [hunk 10]
```diff
@@ -151,93 +128,109 @@ jobs:
...
+      - name: Add armhf architecture
+        run: dpkg --add-architecture armhf
+      - name: Install build deps for armhf
+        run: apt update && apt install -y gcc-arm-linux-gnueabihf g++-arm-linux-gnueabihf libc6-armhf-cross libc6-dev-armhf-cross tar ca-certificates
+      - name: Install ${{ matrix.target }} target
+        run: rustup target add ${{ matrix.target }}
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:141]: I think we can remove this rustup?

File: .github/workflows/docker-build-static.yml [hunk 11]
```diff
@@ -151,93 +128,109 @@ jobs:
...
-          path: target/x86_64-unknown-linux-musl/release/lldap
-      - name: upload amd64 migration-tool artifacts
-        uses: actions/upload-artifact@v3
-        with:
-          name: amd64-migration-tool-bin
-          path: target/x86_64-unknown-linux-musl/release/migration-tool
+       - name: Download all artifacts
+         uses: actions/download-artifact@v3
+         with:
+           path: bin/
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:185]: We only need one artifact, no?

File: .github/workflows/docker-build-static.yml [hunk 12]
```diff
@@ -357,25 +340,25 @@ jobs:
 ### Download artifacts, clean up ui, upload to release page ###
 ###############################################################
   create-release-artifacts:
-     needs: [build-ui,build-armhf,build-aarch64,build-amd64]
+     needs: [lldap-test]
```
Comment (nitnelave) [inline .github/workflows/docker-build-static.yml:343]: Why `lldap-test`? I'd rather just depend on the builds.

**Review by nitnelave — CHANGES_REQUESTED** [review]

File: .github/workflows/docker-build-static.yml (hunk 10 above)
Comment (martadinata666) [reply .github/workflows/docker-build-static.yml:141]: Oh the base image updated, should be removed then.

File: .github/workflows/docker-build-static.yml (hunk 12 above)
Comment (martadinata666) [reply .github/workflows/docker-build-static.yml:343]: Im thinking that that it should pass test first?

File: .github/workflows/Dockerfile.dev (hunk 5 above)
Comment (martadinata666) [reply .github/workflows/Dockerfile.dev]: im updating the dockerfile dev to include rust musl target. ☕

File: .github/workflows/docker-build-static.yml (hunk 12 above)
Comment (nitnelave) [reply .github/workflows/docker-build-static.yml:330]: oh, right, for release it makes sense. Although it's already checked when submitting the PR.

Comment (martadinata666) [reply .github/workflows/docker-build-static.yml]: gonna revert to old behavior then

**Comment by nitnelave** [comment]
FYI, I pushed the rust-dev image with aarch musl

**Comment by nitnelave** [comment]
Btw, you can just ping me when it's ready for review.

**Comment by martadinata666** [comment]
It should be ready, maybe there is some nit you want? ☕

**Review by nitnelave — APPROVED** [review]

**Pull request merged**